### PR TITLE
Add Metadata to further models

### PR DESCRIPTION
**Pull request by miggland**
Adds the metadata mixin, and thus the ability to add metadata to the following models, as mentioned in https://github.com/inventree/InvenTree/issues/4202 (List here is slightly different):

- [X] Manufacturer part
- [X] Supplier part
- [X] Part labels
- [X] Stock item labels
- [X] Stock location labels
- [X] PO lines
- [X] PO extra lines
- [X] SO lines
- [X] SO extra lines
- [X] SO shipments
- [X] Build Orders
- [X] Build order items
- [X] BOM items
- [X] Part parameter templates
- [X] Reports BOM
- [X] Reports Build
- [X] Reports PO
- [X] Reports SO
- [X] Reports stock item test

To do:
- [x] Migrations
- [x] Tests
- [x] Update API version

### CODE REVIEW

**Comment by miggland** [comment]
@SchrodingersGat @matmair: After adding metadata to a BOM, it seems there is an extra column included in the BOM export file, which means this test fails:

```
FAIL: test_export_csv (part.test_bom_export.BomExportTest)
Test BOM download in CSV format.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/inventree/InvenTree/part/test_bom_export.py", line 136, in test_export_csv
    self.assertIn(header, expected)
AssertionError: 'metadata' not found in ['BOM Level', 'BOM Item ID', 'Parent ID', 'Parent IPN', 'Parent Name', 'Part ID', 'Part IPN', 'Part Name', 'Description', 'Assembly', 'Quantity', 'optional', 'consumable', 'overage', 'Reference', 'Note', 'inherited', 'allow_variants', 'Default Location', 'Total Stock', 'Available Stock', 'On Order']
```

What's you take on this - should metadata be included in BOM export or should metadata be excluded from the export somehow?

File: InvenTree/report/api.py [hunk 1]
```diff
@@ -16,11 +16,12 @@
...
 import part.models
-from InvenTree.mixins import ListAPI, RetrieveAPI, RetrieveUpdateDestroyAPI
+import report.models
+from InvenTree.mixins import (ListAPI, RetrieveAPI, RetrieveUpdateAPI,
+                              RetrieveUpdateDestroyAPI)
+from plugin.serializers import MetadataSerializer
 from stock.models import StockItem, StockItemAttachment
 
-from .models import (BillOfMaterialsReport, BuildReport, PurchaseOrderReport,
-                     SalesOrderReport, TestReport)
```
Comment (matmair) [inline InvenTree/report/api.py:23]: Question: Why this rewrite?

File: InvenTree/report/api.py [hunk 2]
```diff
@@ -295,13 +296,13 @@ def filter_queryset(self, queryset):
 
             valid_report_ids = set()
 
-            for report in queryset.all():
+            for qreport in queryset.all():
```
Comment (matmair) [inline InvenTree/report/api.py:299]: Comment: Without changing the imports this variable rename would be unnecesarry

File: InvenTree/report/api.py [hunk 3]
```diff
@@ -683,13 +684,36 @@ def get(self, request, *args, **kwargs):
...
 
 
+class ReportMetadata(RetrieveUpdateAPI):
+    """API endpoint for viewing / updating Report metadata."""
+
+    def get_serializer(self, *args, **kwargs):
+        """Return correct MetadataSerializer instance depending on which model is requested"""
+        # Get type of report, make sure its one of the allowed values
+        reporttype = self.kwargs.get('reporttype', 'PurchaseOrder')
+        assert reporttype in ['PurchaseOrder', 'SalesOrder', 'Build', 'BillOfMaterials', 'Test']
```
Comment (matmair) [inline InvenTree/report/api.py:694]: Please change: asserts are not garunteed to be followed in python, please raise explicitly if the condition is not met.

File: InvenTree/report/api.py [hunk 4]
```diff
@@ -683,13 +684,36 @@ def get(self, request, *args, **kwargs):
...
+        # Return corresponding Serializer
+        UseModel = getattr(report.models, f"{reporttype}Report")
+        return MetadataSerializer(UseModel, *args, **kwargs)
+
+    def get_queryset(self, *args, **kwargs):
+        """Return correct queryset depending on which model is requested"""
+        # Get type of report, make sure its one of the allowed values
+        reporttype = self.kwargs.get('reporttype', 'PurchaseOrder')
+        assert reporttype in ['PurchaseOrder', 'SalesOrder', 'Build', 'BillOfMaterials', 'Test']
```
Comment (matmair) [inline InvenTree/report/api.py:703]: Please change: See above

File: InvenTree/report/api.py [hunk 5]
```diff
@@ -683,13 +684,36 @@ def get(self, request, *args, **kwargs):
         return self.print(request, orders)
 
 
+class ReportMetadata(RetrieveUpdateAPI):
+    """API endpoint for viewing / updating Report metadata."""
+
+    def get_serializer(self, *args, **kwargs):
+        """Return correct MetadataSerializer instance depending on which model is requested"""
+        # Get type of report, make sure its one of the allowed values
+        reporttype = self.kwargs.get('reporttype', 'PurchaseOrder')
```
Comment (matmair) [inline InvenTree/report/api.py:693]: Question: Is it intended that no argument = PO? Is there a need for a default value or could we raise it if no kwarg is defined? To make sure it is better debuggable.

**Review by matmair — CHANGES_REQUESTED** [review]
Looks good in general @miggland! Please take look at the suggestions/change requests.

**Comment by matmair** [comment]
> @SchrodingersGat @matmair: After adding metadata to a BOM, it seems there is an extra column included in the BOM export file, which means this test fails:
> 
> ```
> FAIL: test_export_csv (part.test_bom_export.BomExportTest)
> Test BOM download in CSV format.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/inventree/InvenTree/part/test_bom_export.py", line 136, in test_export_csv
>     self.assertIn(header, expected)
> AssertionError: 'metadata' not found in ['BOM Level', 'BOM Item ID', 'Parent ID', 'Parent IPN', 'Parent Name', 'Part ID', 'Part IPN', 'Part Name', 'Description', 'Assembly', 'Quantity', 'optional', 'consumable', 'overage', 'Reference', 'Note', 'inherited', 'allow_variants', 'Default Location', 'Total Stock', 'Available Stock', 'On Order']
> ```
> 
> What's you take on this - should metadata be included in BOM export or should metadata be excluded from the export somehow?

Metadata should not be included, it is to be used by plugins - not the user.

File: InvenTree/report/api.py (hunk 1 above)
Comment (miggland) [reply InvenTree/report/api.py:23]: The reason is that I made a combined method for all Report metadata - where I  use getattr based on the argument.

`UseModel = getattr(report.models, f"{reporttype}Report")`

This is not possible (without some other, sketchy solutions using `sys`) if the modules are imported in the main namespace.

If you have a suggestion of how to change the `getattr` method, please let me know.

File: InvenTree/report/api.py (hunk 2 above)
Comment (miggland) [reply InvenTree/report/api.py:299]: Correct. But, see above.

File: InvenTree/report/api.py (hunk 1 above)
Comment (matmair) [reply InvenTree/report/api.py:23]: I would be keener on a strongly typed reference. Names might change, this way refactoring will not work automatically. I will type a suggestion up.

Comment (miggland) [reply InvenTree/report/api.py:23]: OK, please share if you have another solution! However, looking at the other imports let's the one I removed stand out as different:
```
import build.models
import common.models
import InvenTree.helpers
import order.models
import part.models
```

**Comment by matmair** [comment]
@miggland I have made a commit to [this branch](https://github.com/matmair/InvenTree/tree/miggland/metadata-hardreferences) that rolls back most of the import changes and uses hot refs for the view

File: InvenTree/report/api.py (hunk 1 above)
Comment (miggland) [reply InvenTree/report/api.py:23]: > @miggland I have made a commit to [this branch](https://github.com/matmair/InvenTree/tree/miggland/metadata-hardreferences) that rolls back most of the import changes and uses hot refs for the view

Ah, good solution! I'll try to get back to this and incorporate your ideas in the next few days.

**Comment by miggland** [comment]
Anybody have an idea what causes this?

```
5.7124s test_package_loading (plugin.test_plugin.RegistryTests)
5.5662s test_printing_endpoints (plugin.base.label.test_label_mixin.LabelMixinTests)
5.3070s test_printing_process (plugin.base.label.test_label_mixin.LabelMixinTests)
Exception ignored in: <function Local.__del__ at 0x7fc079d0ddc0>
Traceback (most recent call last):
  File "/home/inventree/data/env/lib/python3.9/site-packages/asgiref/local.py", line 89, in __del__
    for context_obj in self._context_refs:
  File "/home/inventree/data/env/lib/python3.9/site-packages/asgiref/local.py", line 100, in __getattr__
    with self._thread_lock:
  File "/home/inventree/data/env/lib/python3.9/site-packages/asgiref/local.py", line 100, in __getattr__
    with self._thread_lock:
  File "/home/inventree/data/env/lib/python3.9/site-packages/asgiref/local.py", line 100, in __getattr__
    with self._thread_lock:
  [Previous line repeated 988 more times]
RecursionError: maximum recursion depth exceeded
```

**Comment by matmair** [comment]
Seems like it could be related to plugins/app.ready() - I will take a look in the next days.

**Comment by miggland** [comment]
The test is still failing on my test VM, with the error given above - even though it's working here. Odd.

It is somehow related to `slowtests`- if I disable those (see change below), then the tests run fine.

I'm also seeing extra migration files created constantly, with the following content. Strange, anything I need to worry about?
```
# Generated by Django 3.2.18 on 2023-03-17 11:06

from django.db import migrations


class Migration(migrations.Migration):

    dependencies = [
        ('InvenTree', '0001_initial'),
    ]

    operations = [
    ]
```
```
# Generated by Django 3.2.18 on 2023-03-17 12:00

from django.db import migrations


class Migration(migrations.Migration):

    dependencies = [
        ('InvenTree', '0002_auto_20230317_1106'),
    ]

    operations = [
    ]

```


Change to slowtests:
```
diff --git a/InvenTree/InvenTree/settings.py b/InvenTree/InvenTree/settings.py
index 1aab68d5d..6f8234917 100644
--- a/InvenTree/InvenTree/settings.py
+++ b/InvenTree/InvenTree/settings.py
@@ -38,8 +38,8 @@ if TESTING:
     PASSWORD_HASHERS = ['django.contrib.auth.hashers.MD5PasswordHasher',]
 
     # Enable slow-test-runner
-    TEST_RUNNER = 'django_slowtests.testrunner.DiscoverSlowestTestsRunner'
-    NUM_SLOW_TESTS = 25
+    #TEST_RUNNER = 'django_slowtests.testrunner.DiscoverSlowestTestsRunner'
+    #NUM_SLOW_TESTS = 25
 
     # Note: The following fix is "required" for docker build workflow
     # Note: 2022-12-12 still unsure why...
diff --git a/tasks.py b/tasks.py
index 85501e024..3788b5f29 100644
--- a/tasks.py
+++ b/tasks.py
@@ -559,7 +559,7 @@ def test(c, disable_pty=False):
     pty = not disable_pty
 
     # Run coverage tests
-    manage(c, 'test --slowreport', pty=pty)
+    manage(c, 'test', pty=pty)
 
 
 @task(help={'dev': 'Set up development environment at the end'})
```

File: InvenTree/report/api.py [hunk 6]
```diff
@@ -418,13 +420,39 @@ class SalesOrderReportPrint(SalesOrderReportMixin, ReportPrintMixin, RetrieveAPI
...
+
+        if reportmodel not in [PurchaseOrderReport, SalesOrderReport, BuildReport, BillOfMaterialsReport, TestReport]:
+            raise ValidationError("Invalid report model")
+        return reportmodel
+
+        # Return corresponding Serializer
+    def get_serializer(self, *args, **kwargs):
+        """Return correct MetadataSerializer instance depending on which model is requested"""
+        # Get type of report, make sure its one of the allowed values
+        UseModel = self._get_model(*args, **kwargs)
```
Comment (SchrodingersGat) [inline InvenTree/report/api.py:439]: Can't we use a simpler approach here:

```python
UseModel = self.Meta.model
```

(or similar)?

The endpoint already has either a model or queryset defined, righ?

**Comment by SchrodingersGat** [comment]
@miggland looking good, but see my comment above

@matmair any thoughts on this?

File: InvenTree/report/api.py (hunk 6 above)
Comment (miggland) [reply InvenTree/report/api.py:439]: This question is for @matmair, he helped out and contributed this part of the code. I'm not sure of the implications / the details of your question..

File: InvenTree/report/api.py [hunk 7]
```diff
@@ -292,7 +294,7 @@ class StockItemTestReportPrint(StockItemTestReportMixin, ReportPrintMixin, Retri
     def report_callback(self, item, report, request):
         """Callback to (optionally) save a copy of the generated report"""
 
-        if common.models.InvenTreeSetting.get_setting('REPORT_ATTACH_TEST_REPORT', cache=False):
+        if common.models.InvenTreeSetting.get_setting('REPORT_ATTACH_TEST_REPORT'):
```
Comment (matmair) [inline InvenTree/report/api.py:297]: @miggland Question: why was this changed?

**Review by matmair — APPROVED** [review]
Lgtm.
There is a caching change I am unsure of - the rest looks great!

File: InvenTree/report/api.py (hunk 6 above)
Comment (matmair) [reply InvenTree/report/api.py:439]: @SchrodingersGat not in this case, the model is provided in the url definition as an argument. Reduced the boilerplate as everything is the same for these 4 models. We can type it out if you prefer that. This mainly cuts down on the LoC.

File: InvenTree/report/api.py (hunk 7 above)
Comment (miggland) [reply InvenTree/report/api.py]: Well spotted. That was not intentional, must have occurred in a merge operation. I've changed it back.

**Comment by miggland** [comment]
Fixed as above, and merged latest master into the branch.

**Comment by SchrodingersGat** [comment]
@miggland many thanks for the great work here :)
